x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java

```diff
@@ -328,72 +330,97 @@ private void deleteJobDocuments(ParentTaskAssigningClient parentTaskClient, Stri
-                        executeAsyncWithOrigin(
-                                parentTaskClient.threadPool().getThreadContext(),
-                                ML_ORIGIN,
-                                request,
-                                ActionListener.<AcknowledgedResponse>wrap(
-                                        response -> deleteByQueryExecutor.onResponse(false), // skip DBQ && Alias
-                                        failureHandler),
-                                parentTaskClient.admin().indices()::delete);
-                    }
-                },
-                failure -> {
-                    if (ExceptionsHelper.unwrapCause(failure) instanceof IndexNotFoundException) { // assume the index is already deleted
-                        deleteByQueryExecutor.onResponse(false); // skip DBQ && Alias
-                    } else {
-                        failureHandler.accept(failure);
-                    }
-                }
+        // Step 6. Handle each multi-search response. There should be one response for each underlying index.
+        // For each underlying index that contains results ONLY for the current job, we will delete that index.
+        // If there exists at least 1 index that has another job's results, we will run DBQ.
+        ActionListener<MultiSearchResponse> customIndexSearchHandler = ActionListener.wrap(
+           multiSearchResponse -> {
+               if (multiSearchResponse == null) {
+                   deleteByQueryExecutor.onResponse(true); // We need to run DBQ and alias deletion
+                   return;
+               }
+               String defaultSharedIndex = AnomalyDetectorsIndexFields.RESULTS_INDEX_PREFIX +
+                   AnomalyDetectorsIndexFields.RESULTS_INDEX_DEFAULT;
+               List<String> indicesToDelete = new ArrayList<>();
+               boolean needToRunDBQTemp = false;
+               assert multiSearchResponse.getResponses().length == indexNames.get().length;
+               int i = 0;
+               for (MultiSearchResponse.Item item : multiSearchResponse.getResponses()) {
+                   if (item.isFailure()) {
+                       ++i;
+                       if (ExceptionsHelper.unwrapCause(item.getFailure()) instanceof IndexNotFoundException) {
+                           // index is already deleted, no need to take action against it
+                           continue;
+                       } else {
+                           failureHandler.accept(item.getFailure());
+                           return;
+                       }
+                   }
+                   SearchResponse searchResponse = item.getResponse();
+                   if (searchResponse.getHits().getTotalHits().value > 0 || indexNames.get()[i].equals(defaultSharedIndex)) {
+                       ++i;
+                       needToRunDBQTemp = true;
+                   } else {
+                       indicesToDelete.add(indexNames.get()[i++]);
```

Comment:
You seem to increment `i` in both branches. Could the increment statement be factored out into the line 366?

Reply:
for sure